src/optimistic-oracle/optimistic-oracle.js

```diff
@@ -0,0 +1,232 @@
+  const tokenAddress = contractAddress.toLowerCase();
+
+  // check that the currency denomination we requested is supported
+  if (!(Object.prototype.hasOwnProperty.call(response.data[tokenAddress], vsCurrency))) {
+    throw new Error(`CoinGecko has no data for token '${tokenAddress}' vs currency '${vsCurrency}'`);
+  }
+
+  const price = response.data[tokenAddress][vsCurrency];
+
+  return price;
+}
+
+function provideHandleTransaction(erc20Contract) {
+  return async function handleTransaction(txEvent) {
+    const findings = [];
+
+    if (optimisticOracleAddress === undefined) {
+      // get the Optimistic Oracle contract address for mainnet
+      // the address returned by the promise will be lowercase
+      optimisticOracleAddress = await getAddress('OptimisticOracle', CHAIN_ID);
+    }
+
+    // filter only logs that match the optimistic oracle address
+    // test transaction for ProposePrice event:
+    // npx forta-agent run --tx 0x76ff352b2665886a2a3d3b16fe0fa41f61e4ffc0824b3a6734383d397187f53f
+    const oracleLogs = txEvent.logs.filter((log) => log.address === optimisticOracleAddress);
+
+    if (oracleLogs === []) return findings;
+
+    // parse oracle logs for our target events:  RequestPrice and ProposePrice
+    const parse = (log) => iface.parseLog(log);
+    const filter = (log) => eventNames.indexOf(log.name) !== -1;
+    const parsedLogs = oracleLogs.map(parse).filter(filter);
+
+    // process the target events
+    /* eslint-disable no-await-in-loop */
+    for (let i = 0; i < parsedLogs.length; i++) {
+      const log = parsedLogs[i];
+      if (log.name === 'RequestPrice') {
+        // get token address (this will be checksum encoded)
+        const { currency, requester } = log.args;
+
+        // make a request to the price feed API
+        let price;
+        try {
+          price = await getPrice(currency);
+        } catch (error) {
+          console.error(error);
+          continue;
+        }
+
+        // report the price obtained as a finding
+        findings.push(Finding.fromObject({
+          name: 'UMA Price Request',
+          description: `Price requested from Optimistic Oracle.  Currency=${currency}, Price=${price}`,
+          alertId: 'AE-UMA-OO-REQUESTPRICE',
+          severity: FindingSeverity.Low,
+          type: FindingType.Unknown,
+          protocol: 'uma',
+          everestId: config.umaEverestId,
+          metadata: {
+            requester,
+            currency,
+            price,
+          },
+        }));
+      }
+
+      if (log.name === 'ProposePrice') {
+        const { currency } = log.args;
+        const { requester } = log.args;
+        const { proposer } = log.args;
+
+        // set up ERC-20 contract to get decimals value
+        // in production, erc20Contract will always be undefined
+        // in testing, erc20Contract will be assigned to a mock contract
+        /* eslint-disable no-param-reassign */
+        if (erc20Contract === undefined) {
```

Comment:
We expect this `Contract` object to be set up identically every time.  Perhaps setting up the `Contract` object could be moved to the `provideHandleTransaction` function (after line 109 but before the `handleTransaction` function is returned).

With that move, the check for `undefined` would still occur (albeit once), the `Contract` object would be created and assigned (if necessary), and all subsequent executions of `handleTransaction` would receive the same `Contract ` object, rather than creating their own `Contract` object for every `ProposePrice` event.

Reply:
The `currency` parameter can change on every call, so erc20Contract needs to be created dynamically.

Reply:
Is the following possible?
Two `ProposePrice` events in one transaction, each event using a different currency.